### ISSUE TITLE
Dict access: Inconsistent treatment of Literals

### ISSUE THREAD

**Issue by twoertwein** [microsoft/pyright#1891]
**Describe the bug**
I'm not sure how to summarize it :) I hope the example is clear.

**To Reproduce**
```py
def foo() -> str:
    return "a"

def foo2() -> tuple[str]:
    return ("a",)

# 1
test = {"a": 1}
test[foo()]  # works

# 2
test2 = {("a",): 1}
test2[foo2()]  # fails (mypy passes)

# 3
key = foo2()
if key in test2:
    test2[key]  # fails (mypy passes)

# 4
key_int = 1
if key_int in test2:
    test2[key_int]  # fails (mypy fails)
```
```
error: Argument of type "tuple[str]" cannot be assigned to parameter "k" of type "tuple[Literal['a']]" in function "__getitem__"
    Tuple entry 1 is incorrect type
      "str" cannot be assigned to type "Literal['a']" (reportGeneralTypeIssues)
error: Argument of type "tuple[str]" cannot be assigned to parameter "k" of type "tuple[Literal['a']]" in function "__getitem__"
    Tuple entry 1 is incorrect type
      "str" cannot be assigned to type "Literal['a']" (reportGeneralTypeIssues)
error: Argument of type "Literal[1]" cannot be assigned to parameter "k" of type "tuple[Literal['a']]" in function "__getitem__"
    "Literal[1]" is incompatible with "tuple[Literal['a']]" (reportGeneralTypeIssues)
```

**Expected behavior**
If 1 works, 2 should probably work as well.

I'm tempted to say that 3 should work (key has a literal-compatible type and through the if statement the key could be narrowed down to one of the possible literals).

I'm not sure about 4: it is a type error but the access is guarded. Maybe (an additional) "code unreachable" would make sense?

**Comment by erictraut** [comment]
Type checkers vary in their type inference rules when variables are not provided with explicit type annotations. That explains the difference you're seeing between pyright and mypy in these cases. 

We've spent a lot of time tuning our type inference rules, and the examples you provided are all working as intended. For details in how pyright infers types for dicts, lists, and tuples, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).

The most relevant information here is that pyright infers literal types for tuple expressions, since tuples are immutable. But it does not infer literal types for lists, dicts or sets, because those are mutable containers, and it's common to add more entries to these containers after initial creation.

Type inference rules involve heuristics and assumptions, and they will never be 100% correct regardless of what algorithm is used. In some cases (like some of the examples above), you will need to provide explicit type annotations to guide the type checker. 

For example, if you add the following annotation, your example 2 and 3 will type check without error.

```python
test2: Dict[Tuple[str, ...], int] = {("a",): 1}
```

**Comment by twoertwein** [comment]
That makes sense, thank you very much for the explanation!
